library/Fission/AWS/Types.hs

```diff
@@ -30,6 +31,12 @@ instance MimeRender PlainText DomainName where
 instance MimeRender OctetStream DomainName where
   mimeRender _ = UTF8.textToLazyBS . getDomainName
 
+instance MimeUnrender PlainText DomainName where
+  mimeUnrender _proxy bs =
+    case decodeUtf8' $ Lazy.toStrict bs of
+      Left err  -> Left $ show err
+      Right txt -> Right $ DomainName txt
```

Comment:
This is readable, so purely for educational purposes:

If we didn't have two constructors to get inside of here, we could have used `fmap`, right?

```haskell
fmap DomainName . decodeUtf8' $ Lazy.toStrict bs -- assuming always `Right`
```

There's a class that's like `Functor`, but for this situation: [`Bifunctor`](https://www.stackage.org/haddock/lts-14.8/base-4.12.0.0/Data-Bifunctor.html), which defined [`bimap`](https://www.stackage.org/haddock/lts-14.8/base-4.12.0.0/Data-Bifunctor.html#v:bimap)

```haskell
bimap show DomainName . decodeUtf8' $ Lazy.toStrict bs
```

That does the same as the `case` version. Now to be clear, I use the `case` version essentially always, because it's very easy to see what's happening. Bifunctor

Reply:
oh nice that's cool ^^
Very useful to see it done both ways, but yeah with the `case` version it's definitely more immediately apparent what's going on